component/stream/read_write_filename.go

```diff
@@ -0,0 +1,476 @@
+	err := rw.FlushFile(internal.FlushFileOptions(options))
+	if err != nil {
+		log.Err("Stream::FlushFile : error flushing file %s [%s]", options.Handle.Path, err.Error())
+		return err
+	}
+	if !rw.StreamOnly {
+		rw.purge(options.Handle.Path, true)
+	}
+	err = rw.NextComponent().CloseFile(options)
+	if err != nil {
+		log.Err("Stream::CloseFile : error closing file %s [%s]", options.Handle.Path, err.Error())
+	}
+	return err
+}
+
+func (rw *ReadWriteFilenameCache) FlushFile(options internal.FlushFileOptions) error {
+	// log.Trace("Stream::CloseFile : name=%s, handle=%d", options.Handle.Path, options.Handle.ID)
+	if options.Handle.Dirty() {
+		err := rw.NextComponent().FlushFile(options)
+		if err != nil {
+			log.Err("Stream::FlushFile : error flushing file %s [%s]", options.Handle.Path, err.Error())
+			return err
+		}
+		options.Handle.Flags.Clear(handlemap.HandleFlagDirty)
+	}
+	return nil
+}
+
+func (rw *ReadWriteFilenameCache) DeleteFile(options internal.DeleteFileOptions) error {
+	log.Trace("Stream::DeleteFile : name=%s", options.Name)
+	err := rw.NextComponent().DeleteFile(options)
+	if err != nil {
+		log.Err("Stream::DeleteFile : error deleting file %s [%s]", options.Name, err.Error())
+		return err
+	}
+	if !rw.StreamOnly {
+		rw.purge(options.Name, false)
+	}
+	return nil
+}
+
+func (rw *ReadWriteFilenameCache) DeleteDirectory(options internal.DeleteDirOptions) error {
+	log.Trace("Stream::DeleteDirectory : name=%s", options.Name)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Name) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().DeleteDir(options)
+	if err != nil {
+		log.Err("Stream::DeleteDirectory : error deleting directory %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+func (rw *ReadWriteFilenameCache) RenameDirectory(options internal.RenameDirOptions) error {
+	log.Trace("Stream::RenameDirectory : name=%s", options.Src)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Src) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().RenameDir(options)
+	if err != nil {
+		log.Err("Stream::RenameDirectory : error renaming directory %s [%s]", options.Src, err.Error())
+		return err
+	}
+	return nil
+}
+
+// Stop : Stop the component functionality and kill all threads started
+func (rw *ReadWriteFilenameCache) Stop() error {
+	log.Trace("Stopping component : %s", rw.Name())
+	if !rw.StreamOnly {
+		rw.Lock()
+		defer rw.Unlock()
+		for fileName, buffer := range rw.fileCache {
+			delete(rw.fileCache, fileName)
+			buffer.Lock()
+			defer buffer.Unlock()
+			buffer.Purge()
+			atomic.AddInt32(&rw.CachedObjects, -1)
+		}
+	}
+	return nil
+}
+
+// GetAttr : Try to serve the request from the attribute cache, otherwise cache attributes of the path returned by next component
+func (rw *ReadWriteFilenameCache) GetAttr(options internal.GetAttrOptions) (*internal.ObjAttr, error) {
```

Comment:
Should we also add SyncFile impl? this should also be applied to the other read write impl too right?

Reply:
IsDirEmpty, StreamDir, ReadDir these should be implemented here too right? same for other read write impl

Reply:
I think syncFile migh tneed to be implemented - however the dir ones should be passed to storage. Is there a reason to implement them?

Reply:
In case we need to update the file data - size for example. We wont hit a case where the file exists in the stream cache but isnt in the container right?

Reply:
we won't have a case of file in cache but not in container so "isDirEmpty" should not be implemented. As for the other two - I guess they would be useful but only if the developer wrote to the file (and we updated the LMT in cache only) and then listed. This however would probably only apply to filename caching (where we do update the LMT). But it would not apply to handle level we dont adjust the LMT since getAttr is not on a handle level

Reply:
will add in a separate pr